prow/spyglass/viewers/viewers.go

```
@@ -0,0 +1,201 @@
+}
+
+// ViewHandler consumes artifacts and some possible callback json data and returns an html view.
+// Use the javascript function refreshView(viewName, viewData) to allow your viewer to call back to itself
+// (request more data, update the view, etc.). ViewData is a json blob that will be passed back to the
+// handler function for your view as the string
+type ViewHandler func([]Artifact, string) string
+
+// View gets the updated view from an artifact viewer with the provided name
+func View(name string, artifacts []Artifact, raw string) (string, error) {
+	handler, ok := viewHandlerRegistry[name]
+	if !ok {
+		return "", ErrInvalidViewName
+	}
+	return handler(artifacts, raw), nil
+
+}
+
+// Title gets the title of the view with the given name
+func Title(name string) (string, error) {
+	m, ok := viewMetadataRegistry[name]
+	if !ok {
+		return "", ErrInvalidViewName
+	}
+	return m.Title, nil
+
+}
+
+// Priority gets the priority of the view with the given name
+func Priority(name string) (int, error) {
+	m, ok := viewMetadataRegistry[name]
+	if !ok {
+		return -1, ErrInvalidViewName
+	}
+	return m.Priority, nil
+
+}
+
+// RegisterViewer registers new viewers
+func RegisterViewer(viewerName string, metadata ViewMetadata, handler ViewHandler) error {
+	_, ok := viewHandlerRegistry[viewerName]
+	if ok {
+		return fmt.Errorf("viewer already registered with name %s", viewerName)
+	}
+
+	if metadata.Title == "" {
+		return errors.New("empty title field in view metadata")
+	}
+	if metadata.Priority < 0 {
+		return errors.New("priority must be greater than 0")
+	}
+	viewHandlerRegistry[viewerName] = handler
+	viewMetadataRegistry[viewerName] = metadata
+	logrus.Infof("Spyglass registered viewer %s with title %s.", viewerName, metadata.Title)
+	return nil
+}
+
+// UnregisterViewer unregisters viewers
+func UnregisterViewer(viewerName string) error {
+	_, ok := viewHandlerRegistry[viewerName]
+	if !ok {
+		return ErrInvalidViewName
```

Why do we need to return an error if it isn't registered? `delete(map, missingKey)` will just silently return in this scenario: https://blog.golang.org/go-maps-in-action

---

good point, removed error